### DISCUSSION TITLE
sqlalchemy.exc.PendingRollbackError

### DISCUSSION THREAD

**Discussion by psqjgsw45** [sqlalchemy/sqlalchemy#10240] [Usage Questions]
Now I have encountered a more confused problem
My code is below
SQLAlchemy  1.4.48
```
SQLALCHEMY_DATABASE_URI: str = f"mysql+pymysql://{mysql_user}:{mysql_pwd}@{mysql_ip}:3306/{mysql_database}?charset=utf8"
engine = create_engine(SQLALCHEMY_DATABASE_URI,pool_pre_ping=True, max_overflow=16300, pool_recycle=300)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
session = scoped_session(SessionLocal)
# Sends a heartbeat every 30 seconds like mysql
def task_mysql_keep_alive():
    logger.debug ('task_mysql_keep_alive', datetime.now())
    pool = SessionLocal().bind.pool
    active_connections = pool.checkedin()  
    session.execute(text('SELECT 1'))
    session.close()

```
After about 10 hours or so, once again request interface will be submitted to the pymysql. Err. OperationalError: (2013, 'Lost connection to MySQL server during query ")
sqlalchemy.exc.PendingRollbackError: Can't reconnect until invalid transaction is rolled back. (Background on this error at: https://sqlalche.me/e/14/8s2b)
`@router.get("/test/async")
async def test(session=Depends(get_db)):
    session.query(Organization).filter()
    return dict(code=200, msg='success')`
`def get_db():
    connect_control = settings.CONNECT_CONTROL

    try:

        yield session
    
        session.commit()
    except HTTPException_Error:
        session.rollback()
    except Exception as e:
        logger.exception(e)
        session.rollback()
    finally:
        session.close()
  `

**Comment by zzzeek** [comment]
hi -

difficult to comment here since your message has a few different seemingly unrelated fragments of code (a function for "task_mysql_keep_alive" , then later some kind of `async def`.

A few comments I can make include, the `pymysql` driver doesn't work with asyncio, and it looks like you are using the synchronous SQLAlchemy API inside of async methods, that won't exactly "fail" but it's not how asyncio is meant to be used.


The error for MySQL here can actually mean a lot of things as the driver will raise that error if transactional commands are given out of sync, which can happen for various reasons when handling errors and such.     I would not have any way to begin guessing without a complete stack trace.

**Reply by CaselIT** [reply]
It seems that scoped session is used in async functions. That may explain the error
